### PR TITLE
Add iOS spring animation objc files

### PR DESCRIPTION
**Pull request by luckysmg**
This is the second one of https://github.com/flutter/engine/pull/37604#issuecomment-1376593118.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing tests are passing.

### CODE REVIEW

**Comment by luckysmg** [comment]
Emm not sure how to test this.......

**Comment by Hixie** [comment]
Licensing stuff looks good, I'll defer to someone more familiar with ObjectiveC on the details :-)

File: third_party/spring_animation/spring_animation.mm [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  }
+  return self;
+}
+
+- (double)curveFunction:(double)t {
+  double y;
+  if (self.zeta < 1) {
+    // Under damped
+    double envelope = exp(-self.zeta * self.omega0 * t);
+    y = 1 - envelope * ((self.initialVelocity + self.zeta * self.omega0) / self.omega1 *
```
Comment (cyanglaz) [inline third_party/spring_animation/spring_animation.mm:37]: in the js version, V0 is -self.initialVelocity, should this be:
`(-self.initialVelocity + self.zeta * self.omega0)`

Comment (cyanglaz) [reply third_party/spring_animation/spring_animation.mm]: We have left out lots of stuff from the js file. Is this all we needed for our use case?

Comment (luckysmg) [reply third_party/spring_animation/spring_animation.mm]: Yes， See https://github.com/flutter/engine/pull/37604

**Comment by chinmaygarde** [comment]
@cyanglaz Looks like this is ready for re-review.

**Review by cyanglaz — COMMENTED** [review]
Code looks good, looks like this PR is missing tests?

File: shell/platform/darwin/ios/BUILD.gn [hunk 2]
```diff
@@ -270,6 +270,7 @@ shared_library("ios_test_flutter") {
     "//build/config:symbol_visibility_hidden",
   ]
   sources = [
+    "../../../../third_party/spring_animation/SpringAnimationTest.mm",
```
Comment (cyanglaz) [inline shell/platform/darwin/ios/BUILD.gn:273]: I think tests from 3rd party library should be added in its own BUILD file. I'm not sure if our CI runs all the tests in the 3rd party folders tho, maybe someone from the engine team could explain or provide an example? @chinmaygarde @dnfield

Comment (luckysmg) [reply shell/platform/darwin/ios/BUILD.gn:273]: > I'm not sure if our CI runs all the tests in the 3rd party folders tho

Yes. `testing/run_tests.py --type=objc` will run the `SpringAnimationTest.mm`

Comment (luckysmg) [reply shell/platform/darwin/ios/BUILD.gn]: Added. Plz re-review^_^

You can see new tests in  `Mac Unopt`

**Review by cyanglaz — APPROVED** [review]
LGTM

**Comment by luckysmg** [comment]
@Hixie Would you mind marking your approval if all licensing stuff looks good ^_^ ?

**Pull request merged**